src/sysext/sysext.c

```diff
@@ -534,26 +754,31 @@ static int merge_hierarchy(
         assert(meta_path);
         assert(overlay_path);
 
-        /* Resolve the path of the host's version of the hierarchy, i.e. what we want to use as lowest layer
-         * in the overlayfs stack. */
-        r = chase(hierarchy, arg_root, CHASE_PREFIX_ROOT, &resolved_hierarchy, NULL);
+        f = determine_mutable_layer_path_for_hierarchy(hierarchy);
+        if (!f)
+                return log_oom();
+
+        r = chase(f, arg_root, CHASE_PREFIX_ROOT, &resolved_upper_dir, NULL);
         if (r == -ENOENT)
-                log_debug_errno(r, "Hierarchy '%s' on host doesn't exist, not merging.", hierarchy);
+                log_debug_errno(r, "Hierarchy '%s' on host has no directory %s for storing writes, merging as read-only.", hierarchy, f);
         else if (r < 0)
-                return log_error_errno(r, "Failed to resolve host hierarchy '%s': %m", hierarchy);
-        else {
-                r = dir_is_empty(resolved_hierarchy, /* ignore_hidden_or_backup= */ false);
+                return log_error_errno(r, "Failed to resolve directory for storing writes on host hierarchy '%s': %m", hierarchy);
+
+        _cleanup_free_ char *work_dir = NULL;
+
+        if (resolved_upper_dir) {
+                if (path_is_read_only_fs(resolved_upper_dir))
```

Comment:
this call can fail. error needs to be handled somehow

Reply:
Done.